### ISSUE TITLE
Moving to PCL?

### ISSUE THREAD

**Issue by MihaMarkic** [yuvaltz/Granular#10]
It seems to me that moving to PCL (Common, Presentation) might improve the solution (no linked files, easier to manage, more strict type-won't import anything exotic) and perhaps solve the ICommand interface location.
I did take a quick look and it seems that there is mostly reflection related issue if moving to PCL.

What do you think?

**Comment by yuvaltz** [comment]
I'm not familiar with PCL, it definitely sounds interesting, I'll have to look into it

**Comment by MihaMarkic** [comment]
I can try converting it as well.

**Comment by MihaMarkic** [comment]
@yuvaltz I created Common, Presentation, Presentation.Generic and Hosts.Wpf as PCL. I've postfixed these projects with .Pcl to distinguish them but if this works, then these will be the only assemblies (no more duplication).
The refactored code is not exactly tested, it is a proof of concept at this time. One significant change is that Host has to pass an array of referenced assemblies to PCL (Globals.AllAssemblies property) since it can't be done from within PCL.
I've tried to build such a Granular WPF test client, but build task complains that it can't find Application. Can you check it out?

https://github.com/MihaMarkic/Granular/commits/pcl

**Comment by MihaMarkic** [comment]
The other thing worth noting is that I increased minimum .net version to 4.5. Which makes sense anyway since .net 4.0 doesn't support MVVM very well. (i.e. ICommand).

**Comment by yuvaltz** [comment]
I'm not familiar enough with PCL, is it possible to create a PCL class library that can be compiled with .net and Saltarelle? (Saltarelle produces a differerent dll and a js file)

I've noticed that you've changed implementation of "Compatibility" classes, those classes are not shared and supposed to gap the small differences between .net and Saltarelle environment, do you think we do that in a different way?

**Comment by mycall** [comment]
.NET 4.5.2 will be the minimum version of 4.x that Microsoft will support by Jan 2016.

Also, there appears to be two kinds of PCLs, which I don't fully understand: http://blog.nuget.org/20150729/Introducing-nuget-uwp.html

**Comment by MihaMarkic** [comment]
Yep, Saltarelle support for PCL might be a problem. Will check it out.

**Comment by MihaMarkic** [comment]
Sadly it really [looks like Saltarelle has problems with PCL](http://forums.bridge.net/forum/general/open-discussions/911-support-for-portable-class-libary).
Perhaps you could still merge the PCL modified code but not PCL projects? That way, if Bridge 2 (ex-Saltarelle) adds PCL support the transition is easier? But then satarelle enababled projects won't work due to them being .net 2 and new reflection comes only with .net 4.5. Dang, I didn't see that coming (Saltarelle forcing .net 2.0).
I guess just forget about PCL, at least for now. :(

**Comment by yuvaltz** [comment]
That's sad, PCL could've been nice, but I don't think we must have it for now, since there is only one set of libraries that are exposed to the users, the Saltarelle ones (the libraries who are compiled with .net are used only for debugging the framework outside the browser).
We will probably need it in the future when we will support other kinds hosts.

I just need to add a development branch before I could merge your changes.

**Comment by MihaMarkic** [comment]
I'll think a bit more about PCL though.

**Comment by MihaMarkic** [comment]
Perhaps, we could move non-web projects to PCL and leave web ones as they are. After getting more understanding of the whole solution I think we can't get rid of the separation anyway.
This way the non-web part will be compatible with everything supporting this PCL.

**Comment by yuvaltz** [comment]
Sure, we can do that
(we can also wait till we'll have a least one more host (Gtk# maybe), or Saltarelle/Bridge support).

**Comment by MihaMarkic** [comment]
Shall I create a pull request and you can merge it somewhere? If the PCL is the way to go then it might be better sooner than later.

**Comment by yuvaltz** [comment]
I agree, if you could do that it would be great, thanks!

**Comment by MihaMarkic** [comment]
Right now the PCL projects are separate. Do you think I should leave them separate or replace existing ones?

**Comment by yuvaltz** [comment]
You can replace the existing ones
